pennylane/tape/tapes/jacobian_tape.py

```diff
@@ -546,3 +569,163 @@ def jacobian(self, device, params=None, **options):
+
+        By default, the Hessian will be computed with respect to all parameters on the quantum tape.
+        This can be modified by setting the :attr:`~.trainable_params` attribute of the tape.
+
+        The Hessian can be currently computed using only the ``'analytic'`` method.
+
+        Args:
+            device (.Device, .QubitDevice): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameter values are used (via :meth:`~.get_parameters`).
+
+        Keyword Args:
+            method="analytic" (str): The differentiation method. Currently only
+                supports ``"analytic"``.
+            shift=pi/2 (float): the size of the shift for two-term parameter-shift gradient computations
+
+        Returns:
+            array[float]: 2-dimensional array of shape ``(tape.num_params, tape.num_params)``
+
+        **Example**
+
+        .. code-block:: python
+
+            n_wires = 5
+            weights = [2.73943676, 0.16289932, 3.4536312 , 2.73521126, 2.6412488]
+
+            with QubitParamShiftTape() as tape:
+                for i in range(n_wires):
+                    qml.RX(weights[i], wires=i)
+
+                qml.CNOT(wires=[0, 1])
+                qml.CNOT(wires=[2, 1])
+                qml.CNOT(wires=[3, 1])
+                qml.CNOT(wires=[4, 3])
+
+                qml.expval(qml.PauliZ(1))
+
+        If parameters are not provided, the existing tape parameters are used:
+
+        >>> dev = qml.device("default.qubit", wires=n_wires)
+        >>> tape.hessian(dev)
+        array([[ 0.79380556,  0.05549219,  0.10891309, -0.1452963,   0.],
+               [ 0.05549219,  0.79380556, -0.04208544,  0.05614438,  0.],
+               [ 0.10891309, -0.04208544,  0.79380556,  0.11019314,  0.],
+               [-0.1452963,   0.05614438,  0.11019314,  0.79380556,  0.],
+               [ 0.,          0.,          0.,          0.,          0.]])
+
+        Parameters can be optionally passed during execution:
+
+        >>> tape.hessian(dev, params=[1.0, 1.0, 2.0, 0, 0])
+        array([[ 0.12148432, -0.29466251,  0.41341091,  0.,          0.],
+               [-0.29466251,  0.12148432,  0.41341091,  0.,          0.],
+               [ 0.41341091,  0.41341091,  0.12148432,  0.,          0.],
+               [ 0.,          0.,          0.,          0.12148432,  0.],
+               [ 0.,          0.,          0.,          0.,          0.]])
+
+        Parameters provided for execution are temporary, and do not affect
+        the tapes' parameters in-place:
+
+        >>> tape.get_parameters()
+        [2.73943676, 0.16289932, 3.4536312, 2.73521126, 2.6412488]
+
+        If a tape has no trainable parameters, the Hessian will be empty:
+
+        >>> tape.trainable_params = {}
+        >>> tape.hessian(dev)
+        array([], shape=(1, 0), dtype=float64)
```

Comment:
What should we return when there are no trainable parameters? Right now I return a `array([], shape=(1, 0), dtype=float64)`, but I'm not sure what the best practice is here.

Reply:
That's a good question. We should match what the `jacobian()` method does, but I just had a look, and suspect that it might be ambiguous (sometimes it might return `None`, sometimes a zeros array of size `(qnode_output, params)`).

Reply:
Reminder to myself that we should double check that the `jacobian()` method (and the `hessian` method) have consistent output for no trainable parameters